### PR TITLE
neighbors: add cluster_tile APIs and pair outputs

### PR DESCRIPTION
**Pull request by zubatyuk**
# ALCHEMI Toolkit-Ops Pull Request

## Description

Add the cluster-tile neighbor-list strategy and extend neighbor-list outputs across Warp, PyTorch, and JAX: per-pair distances/vectors on demand, inline Warp `pair_fn` evaluation (per-pair energy/force), partial neighbor lists (`target_indices`), selective batched rebuild (`rebuild_flags`), and differentiable Torch/JAX geometry outputs. The core `nvalchemiops.neighbors` package is restructured from flat modules into per-strategy subpackages (`naive/`, `cell_list/`, `cluster_tile/`, `rebuild/`).

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Performance improvement
- [x] Documentation update
- [x] Refactoring (no functional changes)
- [ ] CI/CD or infrastructure change

Not breaking vs 0.3.1: no public 0.3.1 symbol is removed. Retired flat modules (`batch_naive`, `batch_cell_list`, `naive_dual_cutoff`, `batch_naive_dual_cutoff`, `rebuild_detection`) re-export their full 0.3.1 surface with a `DeprecationWarning`; `naive` / `cell_list` are now canonical subpackages re-exporting their 0.3.1 names; `zero_array` still works (now warns and forwards to `array.zero_()`). The one new caller-visible behavior is the method-dependent COO return arity under cell-less `method=None` (see Dispatch Behavior).

## Related Issues

Built on top of #84. Baselines: 0.3.1 (tag `v0.3.1`) has no tile path (only `naive` / `cell_list` with batched and dual-cutoff variants); PR #84 introduced `tile_warp` and rewrote the cell-list / naive launchers. cluster_tile is new since 0.3.1, so the bug fixes below originate in PR #84.

## Changes Made

- Rename the PR #84 tile surface (`tile_warp` / `batch_tile_warp`) to `cluster_tile` / `batch_cluster_tile`; expose single and batched cluster-tile paths through the Warp, PyTorch, and JAX neighbor APIs.
- Restructure core Warp launchers into `naive`, `cell_list`, `cluster_tile`, and `rebuild` subpackages (`*/launchers.py`, `*/dispatch.py`, `*/kernels.py`); flat modules are removed or kept as deprecation shims.
- Add `return_distances` / `return_vectors` to return per-pair scalar distances and displacement vectors alongside matrix or COO neighbor outputs.
- Add inline `pair_fn`: single-cutoff kernels accept a Warp `pair_fn` with `pair_params` plus optional `pair_energies` / `pair_forces` buffers (auto-allocated and returned, matrix- or COO-shaped), with shared argument validation.
- Add partial / selective construction: `target_indices` (partial NL) and `rebuild_flags` (per-system batched rebuild) preserve previous outputs for rows / systems that are not rebuilt.
- Add differentiable geometry: Torch autograd and JAX VJP reconstruct per-pair displacements from emitted indices + shifts in the backward pass; tests cover first- and second-order gradients across naive, cell-list, and cluster-tile bindings where supported.
- Change auto-dispatch to a geometry cost model (atom counts, effective volume, cutoff) instead of the 0.3.1 / PR #84 average-atoms threshold; expose public helpers `estimate_neighbor_list_costs` and `suggest_neighbor_list_method` (Warp, Torch, JAX).
- Deprecate `zero_array`; remove unused `make_outer_neigh_offsets`; bump `warp-lang` to `>= 1.13` (lock `1.11.0` → `1.13.0`); small formatting cleanup in `segment_ops.py`.

## Bug Fixes vs PR #84

Correctness fixes in PR #84's tile path (carried into the rename) and its `batch_cell_list`. Note for benchmarking: a fix that adds previously skipped work makes PR #84 look faster — that is a correctness cost, not a regression.

- Full-fill, was half-fill: the tile query emitted each pair once (`i_sorted < j_sorted`) despite the full-fill contract, undercounting ~2× vs `cell_list`; it now writes both `(i,j)` and `(j,i)` (negated shift), matching `cell_list(half_fill=False)` for neighbor sets, shifts, and autograd forces/energy.
- Tile-buffer overflow → raise + auto-size: the build counted tiles unclamped but clamped writes, silently dropping pairs on dense / high-cutoff systems; it now raises `NeighborOverflowError`, and the default tile buffer is geometry-sized from cutoff + cell density (Torch and JAX, single + batch).
- Dual-cutoff shell miss: tiles were built with the smaller `cutoff`, missing `(cutoff, cutoff2]` pairs; tiles are now built with `max(cutoff, cutoff2)`.
- `batch_cell_list` high-cutoff matrix miscount (fixed by the cell-list flat→package refactor): mis-attributed neighbors across batched rows on large high-cutoff batches; now matches `batch_naive` to the float32 cutoff boundary.
- Launch scaling: query launches over the host-synced emitted-tile count instead of the full allocated buffer, restoring per-atom scaling; consequence — the matrix path is intentionally not `torch.compile` / `jit` graph-capturable (host sync).

## Dispatch Behavior vs 0.3.1 / PR #84

- Auto-selection: `method=None` compares estimated `naive` vs `cell_list` work from per-system atom counts, cutoff, and effective volume (provided cell volume, or a synthesized non-periodic bounding box), replacing the `avg_atoms >= 2000` rule. cluster_tile is also a candidate and is auto-selected when feasible (CUDA, float32, fully-periodic, cell present, large enough, compatible output options) and cheapest; pass an explicit `method=` to force or override.
- Single-system batches: batch metadata describing exactly one system uses the unbatched path (squeezed `(1,3,3)` cell / `(1,3)` pbc), avoiding `batch_cell_list` overhead.
- `batch_cell_list` sub-strategy: the pair-centric query path now requires `avg_atoms_per_system >= 4096` in addition to the cutoff / total-atoms checks, so high-cutoff batches of many small structures stay atom-centric.
- Cell-less COO return arity is method-dependent: with `method=None`, `return_neighbor_list=True`, and no input `cell`, `naive` returns a 2-tuple `(list, ptr)` while `cell_list` / `batch_cell_list` synthesize a non-PBC cell and return a 3-tuple `(list, ptr, shifts)` with zeroed shifts. Pass an explicit `cell`+`pbc` (or use matrix output) for a stable 3-tuple.

## Gaps / Not Implemented

- Kernel factory (`factory.py` / `make_nl_kernel`): not built — per-strategy `get_*_kernel` getters with `lru_cache` provide the specialization and caching a factory would have.
- cluster_tile limitations: no `target_indices`; no `half_fill` (full-fill only); no pair outputs while selectively rebuilding; CUDA + float32 + fully-periodic + cell required.
- JAX cluster_tile: float32 only, eager `cutoff` only. Matrix and COO pair outputs are both supported (COO is eager-only — the matrix→COO pack has a data-dependent pair count; a traced call raises, use `format="matrix"`).
- `pair_fn` energy/force are forward-only on both backends (Warp kernels are `enable_backward=False`); distances/vectors are differentiable to all orders.

## GPU Synchronization

Most NL paths launch fully asynchronously. Device→host synchronizations occur in: cluster_tile / batch_cluster_tile (host-reads emitted tile/pair counts to tighten the launch and to raise `NeighborOverflowError` — the matrix path is intentionally not CUDA-graph / `torch.compile`-capturable); `method=None` auto-dispatch (cost-model read, plus `pbc.all()` for the cluster-tile feasibility check); batched `naive` / `cell_list` sizing (max atoms per system, cell counts, search radii) unless precomputed sizes are passed; single `cell_list` (the three per-axis search-radius components). No sync: single-system `naive` forward, and all JAX traced (`jit` / `grad`) paths (static `ngroup²` tile cap, since `num_tiles` cannot be host-read under trace). For repeated calls, call `suggest_neighbor_list_method` / `estimate_neighbor_list_costs` once on geometry and pass the chosen name as explicit `method=`.

## Testing

- [x] Unit tests pass locally (`make pytest`)
- [x] Linting passes (`make lint`)
- [x] New tests added for new functionality meets coverage expectations? — cluster-tile parity vs `cell_list`, `pair_fn` energy/force vs analytic reference, return_distances/vectors, partial NL, selective rebuild, overflow errors, first- and second-order autograd, and cost-model dispatch.

## Checklist

- [x] I have read and understand the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] I have updated the [CHANGELOG.md](../CHANGELOG.md)
- [x] I have performed a self-review of my code
- [x] I have added docstrings to new functions/classes
- [x] I have updated the documentation (if applicable)

## Additional Notes

JAX status: the traced cluster-tile path uses a trace-safe `ngroup²` upper-triangular tile cap (correct, never overflows, not size-optimized) and is not tight-launched, because the tile-buffer shape must be static under trace; for `ngroup ≤ 256` the capacity equals the old default. `pair_fn` is wired through all five JAX paths (`jax_kernel` for naive / cell-list, `jax_callable` for the tile paths), with energy/force auto-allocated and returned, forward-only, eager-cutoff. JAX `cell_list` / `batch_cell_list` also support `target_indices` + pair outputs and the `pair_centric` strategy (identical pair set to `atom_centric`); only `pair_centric` + `target_indices` is rejected (identical results via `atom_centric`).

> [!TIP]
> This repository uses Greptile, an AI code review service, to help conduct
> pull request reviews. We encourage contributors to read and consider suggestions
> made by Greptile, but note that human maintainers will provide the necessary
> reviews for merging: Greptile's comments are **not** a qualitative judgement
> of your code, nor is it an indication that the PR will be accepted/rejected.
> We encourage the use of emoji reactions to Greptile comments, depending on
> their usefulness and accuracy.

### CODE REVIEW

File: nvalchemiops/torch/neighbors/__init__.py [hunk 1]
```diff
@@ -376,6 +480,45 @@ def neighbor_list(
                 return_neighbor_list=return_neighbor_list,
                 **kwargs,
             )
+        case "cluster_tile":
+            # format="tile" is reachable only via cluster_tile_neighbor_list directly.
+            if cell is None:
+                positions, cell, pbc = synthesize_cell_for_ss(positions, cutoff)
+                pbc = torch.ones(3, dtype=torch.bool, device=positions.device)
+            _reject_unsupported_cluster_tile_combo(pbc, half_fill)
+            return cluster_tile_neighbor_list(
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/torch/neighbors/__init__.py:479]: Fixed in 9549b771dca7d2a746aa57a938d8f479970602df.

File: nvalchemiops/torch/neighbors/__init__.py [hunk 2]
```diff
@@ -376,6 +480,45 @@ def neighbor_list(
...
+                cell,
+                batch_ptr,
+                fill_value=fill_value,
+                format="coo" if return_neighbor_list else "matrix",
+                cutoff2=cutoff2,
+                **kwargs,
+            )
         case "naive_dual_cutoff":
             return naive_neighbor_list_dual_cutoff(
                 positions,
```
[the remark this answers is not in the document]
Comment (zubatyuk) [reply nvalchemiops/torch/neighbors/__init__.py:511]: Fixed in 9549b771dca7d2a746aa57a938d8f479970602df.

File: examples/neighbors/01_simple_neighbor_list.py [hunk 3]
```diff
@@ -83,6 +84,27 @@ def _timed_loop(fn, n_iter=10, warmup=10):
...
+        cell,
+        pbc,
+        cutoff,
+        positions_dtype=positions.dtype,
+        **kwargs,
+    )
+    print(f"\n{label}")
+    for strategy, cost in report:
+        print(f"  {strategy:24s} estimated cost (arbitrary units): {cost:.3g}")
+    return report[0][0]
```
Comment (dallasfoster) [inline examples/neighbors/01_simple_neighbor_list.py]: this method is used twice in this example and I think it might actually make the example more clear during reading if we don't use it.

**Comment by dallasfoster** [comment]
/ok to test [7081f72](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/88/commits/7081f7280839ba714bfb01baa84b7f2b95731f5b)

**Comment by dallasfoster** [comment]
/ok to test [d4bc141](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/88/commits/d4bc1414bb0e230fcc21b5bc6c4593de8662ac17)

**Comment by dallasfoster** [comment]
/ok to test [48ec2c9](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/88/commits/48ec2c9ee10b8df31332d3b5c1aeef2abab8262c)

**Comment by zubatyuk** [comment]
@greptile review

**Comment by zubatyuk** [comment]
@Greptile full review

**Comment by dallasfoster** [comment]
/ok to test [967cabf](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/88/commits/967cabf501a1fb24aecce4392f0b3ca81fb1434f)

**Comment by dallasfoster** [comment]
/ok to test [e0919d0](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/88/commits/e0919d057279b8e391a2ea53dd9a8a7f7e435137)

**Comment by dallasfoster** [comment]
/ok to test [f583045](https://github.com/NVIDIA/nvalchemi-toolkit-ops/pull/88/commits/f58304538cd99c534693cb503d362483c8f107f1)
